jujuclient/utils.go

```diff
@@ -0,0 +1,31 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package jujuclient
+
+import (
+	"github.com/juju/errors"
+)
+
+func updateControllerModels(storedModels *ControllerModels, newModels map[string]ModelDetails) (bool, error) {
+	if len(storedModels.Models) == 0 && len(newModels) == 0 {
+		// Special case: no update necessary.
+		return false, nil
+	}
+	// Add or update controller models based on a new collection.
+	for modelName, details := range newModels {
+		if err := ValidateModel(modelName, details); err != nil {
+			// TODO what to do here?... don't want to stop
+			// an update if one name is invalid..
+			return false, errors.Trace(err)
+		}
+		storedModels.Models[modelName] = details
+	}
+	// Delete models that are not in the new collection.
+	for modelName, _ := range storedModels.Models {
+		if _, ok := newModels[modelName]; !ok {
+			delete(storedModels.Models, modelName)
+		}
+	}
+	return true, nil
```

Comment:
the special case can be dropped by keeping a "changed" variable, making the code a little cleaner; and the function should return false if the model details don't change

```go
changed := len(storedModels.Models) != len(newModels)
for modelName := range storedModels.Models {
    if _, ok := newModels[modelName]; !ok {
        delete(storedModels.Models, modelName)
    }
}
for modelName, details := range newModels {
    oldDetails, ok := storedModels.Models[modelName]
    if ok && details == oldDetails {
        continue
    }
    storedModels.Models[modelName] = details
    changed = true
}
return changed, nil
```

please inline this method in the file store implementation's SetModels. The mem store implementation can just replace the map contents, since it's cheap